package.json

```diff
@@ -58,7 +50,8 @@
     "test-coverage": "node ./out/test/runTest.js --coverage",
     "prettier:write": "prettier --write \"src/**/*.{ts, js}\"",
     "prettier:verify": "prettier --list-different \"src/**/*.{ts, js}\"",
-    "vscode:prepublish": "npm run pretest && npm run test"
+    "bundle:extension": "esbuild ./src/extension.ts  --bundle --outdir=out --format=cjs --target=es2020 --platform=node --external:vscode --external:@salesforce/core --external:@salesforce/lwc-dev-mobile-core --external:@salesforce/sf-plugins-core --minify",
```

Comment:
I'd like to see some changes here. Gonna mention some files outside of this one too.

- Tack to `--outfile` from `--outdir`. The latter is if you need to call multiple targets within a package, which we don't.
- In my experience, the only two `--external:` directives we need are `vscode` and `@oclif/core`: the former because it won't get resolved, and the latter because it tries to access a relative-pathed `package.json` that doesn't otherwise exist in the right place, when the package is inlined into `extension.js`. The other `--external:` directives can go away, as we're including all of `node_modules/`—or rather, _not excluding_ it—in `.vscodeignore`.
- Speaking of `.vscodeignore`: we should ignore all of `out/`, and bang-include (whitelist) `!out/external.js`. We don't need the other artifacts in there—`esbuild` will incorporate them into the external.js end product. And the leftover code could negatively impact the bundle at runtime in unforeseen ways.

Reply:
Cool, let me try and update this.

Reply:
I also updated the build so that it runs a clean, so the out directory SHOULD BE empty. However, perhaps vscode instance could do a concurrent compile and shove *.js files in there too. I'll try the `!out/extension.js` approach.

Reply:
If I change the `--external:*` paths to just have vscode and @oclif/core, the result extension.js bundle is 2.4megs. Leaving in what I have, it is only 10k. I am trying to figure out what exactly it is bundling.

Reply:
If I remove the `--external:*` options (except for vscode and @oclif/core, and also ignore node_modules/*, then the extension does not work at all. It will install, but immediately shows error when launched.

I am still not sure WHY removing these `--external:*` options makes the `out/extension.js` file be so much larger - presumably it is trying to bundle at least PARTS of the dependencies?

Reply:
It does appear that it is attempting to bundle some things in node_modules: `grep ^// extension.js` shows the file paths such as: 

```
// node_modules/@salesforce/ts-types/lib/narrowing/is.js
// node_modules/@salesforce/ts-types/lib/narrowing/as.js
// node_modules/@salesforce/ts-types/lib/errors.js
// node_modules/@salesforce/ts-types/lib/narrowing/to.js
// node_modules/@salesforce/ts-types/lib/narrowing/assert.js
// node_modules/@salesforce/ts-types/lib/narrowing/coerce.js
// node_modules/@salesforce/ts-types/lib/narrowing/ensure.js
// node_modules/@salesforce/ts-types/lib/narrowing/has.js
...
```

Reply:
> and also ignore node_modules/*

We can't ignore `node_modules/` in this iteration. That requires more fine tuning of `--external:` directives than we're going to bite off in V1, to ensure that all of the dependencies of any `--external:` directives are _also_ external. I started playing with that, but it's painful and probably needs automation to be maintainable.

But in the absence of adding any `.vscodeignore` directives for `node_modules`, you should be able to attenuate your `--external:` directives to the two I listed.

> I am still not sure WHY removing these --external:* options makes the out/extension.js file be so much larger - presumably it is trying to bundle at least PARTS of the dependencies?

Correct. When packages are not otherwise called out with `--external:` directives, `esbuild` interrogates and inlines all of those packages and their dependencies _into_ `extension.js`. That's why that file gets bigger—otherwise, `--external:` packages just continue to live on their own in `node_modules/`, and `extension.js` refers to them in a more normal `require` type of way in those cases.

Reply:
> I also updated the build so that it runs a clean, so the out directory SHOULD BE empty. However, perhaps vscode instance could do a concurrent compile and shove *.js files in there too. I'll try the !out/extension.js approach.

Oh that's a good point: `esbuild` may _only_ generate `extension.js`, at which point cleaning beforehand effectively gets us to the same place. I was thinking `esbuild` would generate the same output as e.g. `tsc` does for us in development.